File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    time = 0
    bridge = deque([0 for _ in range(bridge_length)])
    q = deque(truck_weights)
    weight_sum = 0
    
    while q:
        time += 1
        
        weight_sum -= bridge.popleft()
        
        if weight_sum + q[0] <= weight:
            weight_sum += q[0]
            bridge.append(q.popleft())
        else:
            bridge.append(0)
            
    time += bridge_length
    
    return time

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Queue<Integer> q = new LinkedList<>();
        int time = 0;
        int weight_now = 0;
        
        for (int truck : truck_weights) {
            while (true) {
                if (q.isEmpty()) {
                    q.offer(truck);
                    weight_now += truck;
                    time++;
                    break;
                }
                else if (bridge_length == q.size()) {
                    weight_now -= q.poll();
                }
                else {
                    if (weight_now + truck <= weight) {
                        q.offer(truck);
                        weight_now += truck;
                        time++;
                        break;
                    }
                    else {
                        q.offer(0);
                        time++;
                    }
                }
            }
        }
        
        return time + bridge_length;
    }
}